mantidproject/mantid
# Improvements to SimulatedDensityOfStates: force constants import

**Description of work.**
- Implements an alternative file import option to the SimulatedDensityOfStates algorithm, using the Euphonic python package developed as part of the PACE project at ISIS.

  This imports files with "force constants" data, from which frequencies/eigenvectors are computed at arbitrary q-points. This allows the analysis process to be further decoupled from the *ab initio* calculations, and provides access to data computed with the popular Phonopy code. This supports a much wider range of atomistic calculators; up to this point SimulatedDensityOfStates has only been compatible with CASTEP.

- The selection of appropriate q-point sampling meshes can be unintuitive as they are related to the reciprocal lattice vectors. We implement a one-parameter scheme (Moreno-Soler length cutoff) which allows a reasonable default behaviour over many input systems.

- Modifies the Indirect/Simulation Interface to accept the new file formats and dispatch appropriately

This is related to work on the Abins algorithm, as we intend to also provide Euphonic-based inputs there.
As an external Python package with a compiled-C element, including the Euphonic dependency is not entirely trivial. Given that a move to Conda packaging is imminent, it was decided not to add Euphonic to all the existing Mantid build processes. Effectively there are three options available to users:

1. Use a Conda build of Mantid. The Euphonic conda-forge package has been added to the Mantid build recipes, so for Mantid build this way the new features should "just work".
2. Where Mantid was build against a user-accessible Python, `pip install euphonic[phonopy_reader]` should do the trick. It would be helpful if we could arrange for iDAaaS environments to be configured this way.
3. Where Mantid has its own Python environment, Pip needs to be called from inside that environment. A script has been created in the Script Repository which will manage this and refresh the import paths, allowing the new features to be used without restarting mantid. This script is */user/AdamJackson/install_euphonic.py*.

If a user tries to import force constants data using the DensityOfStates interface, an error is produced in the same way as if they try to import IR/Raman data from a .phonon file which is missing that data. A warning message is also produced in the log which directs them to the Script Repository installer.

**To test:**


- Some sample data can be found in the Euphonic test suite: I suggest the *quartz-666-grid.phonon* and *quartz.castep_bin* files from https://github.com/pace-neutrons/Euphonic/tree/master/tests_and_analysis/test/data/castep_files/quartz and *mp-7041-20180417.yaml* file from https://github.com/pace-neutrons/Euphonic/tree/master/tests_and_analysis/test/data/phonopy_files/CaHgO2 . Also grab a benzene example from existing Mantid test data: Data/UnitTest/benzene_Abins.phonon (MD5 ec53aa8ff74b7cd257d7a82c3d6547fd).

*Previously I suggested the CaHgO2-666-grid.yaml file, but this is not a supported file as it does not provide force constants.*

- Make sure that things fail gracefully (and no more than necessary) if the extra library was not installed.
  - On a clean non-Conda installation, open the Indirect/Simulation interface and navigate to DensityOfStates tab. You should be able to select the .phonon file and choose DOS or IR outputs. An error should be raised and handled gracefully if Raman is selected for bezene_Abins.phonon (as that data is missing); CaHgO2-666-grid.yaml is missing both IR and Raman.
  - Try loading some force constants data: either a .castep_bin or .yaml file. This should raise an error, handle it reasonable gracefully, and log an informative message about needing the Euphonic library.
- Install Euphonic with the Script Repository installer, and repeat the above. Hopefully, this time you can get a DensityOfStates using the .castep_bin and .yaml files, and .phonon files continue working as before.
- With a fresh Conda build of Mantid, Euphonic should be available out-of-the-box. You can check this with `import euphonic` from the Workbench IPython interpreter. To check that the C extension is available you could also `import euphonic._euphonic`.


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
